server/src/main/java/org/elasticsearch/monitor/fs/FsHealthService.java

```diff
@@ -0,0 +1,180 @@
+        Setting.timeSetting("monitor.fs.health.slow_path_logging_threshold", TimeValue.timeValueSeconds(5), TimeValue.timeValueMillis(1),
+            Setting.Property.NodeScope, Setting.Property.Dynamic);
+
+
+    public FsHealthService(Settings settings, ClusterSettings clusterSettings, ThreadPool threadPool, NodeEnvironment nodeEnv) {
+        this.threadPool = threadPool;
+        this.enabled = ENABLED_SETTING.get(settings);
+        this.refreshInterval = REFRESH_INTERVAL_SETTING.get(settings);
+        this.slowPathLoggingThreshold = SLOW_PATH_LOGGING_THRESHOLD_SETTING.get(settings);
+        this.currentTimeMillisSupplier = threadPool::relativeTimeInMillis;
+        this.nodeEnv = nodeEnv;
+        this.pathHealthStats = new HashMap<>(1);
+        clusterSettings.addSettingsUpdateConsumer(SLOW_PATH_LOGGING_THRESHOLD_SETTING, this::setSlowPathLoggingThreshold);
+        clusterSettings.addSettingsUpdateConsumer(ENABLED_SETTING, this::setEnabled);
+    }
+
+    @Override
+    protected void doStart() {
+        scheduledFuture = threadPool.scheduleWithFixedDelay(new FsHealthMonitor(), refreshInterval,
+                ThreadPool.Names.GENERIC);
+    }
+
+    @Override
+    protected void doStop() {
+        scheduledFuture.cancel();
+    }
+
+    @Override
+    protected void doClose() throws IOException {
+
+    }
+
+    public void setEnabled(boolean enabled) { this.enabled = enabled; }
+
+    public void setSlowPathLoggingThreshold(TimeValue slowPathLoggingThreshold) {
+        this.slowPathLoggingThreshold = slowPathLoggingThreshold;
+    }
+
+    @Override
+    public StatusInfo getHealth() {
+        StatusInfo statusInfo;
+        if (enabled == false) {
+            statusInfo = new StatusInfo(HEALTHY, "health check disabled");
+        }
+        else if (pathHealthStats.entrySet().stream().anyMatch(map -> map.getValue() == UNHEALTHY)) {
+            String info = "health check failed on [" + pathHealthStats.entrySet()
+                .stream().filter(v -> v.getValue() == UNHEALTHY)
+                .map(k -> k.getKey().toString())
+                .collect(Collectors.joining(",")) + "]";
+
+            statusInfo = new StatusInfo(UNHEALTHY, info);
+        }
+        else {
+            statusInfo = new StatusInfo(HEALTHY, "health check passed");
+        }
+        return statusInfo;
+    }
+
+     class FsHealthMonitor implements Runnable {
+
+        static final String TEMP_FILE_NAME = ".es_temp_file";
+        private byte[] byteToWrite;
+
+        FsHealthMonitor(){
+            this.byteToWrite = new byte[20];
+        }
+
+        @Override
+        public void run() {
+            try {
+                if (enabled) {
+                    monitorFSHealth();
+                    logger.debug("health check succeeded: {}", pathHealthStats);
+                }
+            } catch (Exception e) {
+                logger.error("health check failed", e);
```

Comment:
Can this happen? Looks to me that we already catch and log all exceptions in `monitorFSHealth`.

Reply:
Can happen if `nodeEnv.nodeDataPaths()` fails with lock assertions

Reply:
Ohh yes so it can, thanks. For now this is fine but I'll follow up internally to discuss whether we should consider a failure here to be fatal too.